File: telerik_ui_for_winforms_courseware_chm/Courseware/Grid/CS/RadGridView/CustomCells/ColorSwatchCellElement.cs
using System;
using System.Drawing;
using System.Windows.Forms;
using Telerik.WinControls;
using Telerik.WinControls.UI;

namespace _15_CustomCells
{
    public class ColorSwatchCellElement : GridDataCellElement
    {
        public ColorSwatchCellElement(GridViewColumn column, GridRowElement row) :
        base(column, row)
        {
        }
       
        private LightVisualElement _swatch;

        protected override void CreateChildElements()
        {
            base.CreateChildElements();

            _swatch = new LightVisualElement();
            _swatch.DrawFill = true;
            _swatch.Font = new Font("Arial", 7F, FontStyle.Regular);
            _swatch.Shape = new RoundRectShape(5);
            _swatch.GradientStyle = GradientStyles.Linear;
            _swatch.NumberOfColors = 2;
            _swatch.MouseDown -= new MouseEventHandler(_swatch_MouseDown);
            _swatch.MouseDown += new MouseEventHandler(_swatch_MouseDown);
            this.Children.Add(_swatch);
        }

       
        protected override void SetContentCore(object value)
        {
            // blank out the cell's text, only show on swatch element
            this.Text = "";
        }

        protected override SizeF ArrangeOverride(SizeF finalSize)
        {
            base.ArrangeOverride(finalSize);
            //size the element according to the cell size
            _swatch.Arrange(new RectangleF(3, 3, finalSize.Width - 6, finalSize.Height - 6));
            return finalSize;
        }

        public override void SetContent()
        {
            base.SetContent();

            // set default properties
            _swatch.BackColor = Color.White;
            _swatch.BackColor2 = Color.White;
            _swatch.Text = String.Empty;

            // only process if there is a value and its not "Transparent"
            if (!this.Value.Equals(System.DBNull.Value) &&
                (!this.Value.ToString().Equals("Transparent")))
            {
                string colorName = this.Value.ToString();
                // convert to a Color type and assign to background color
                _swatch.BackColor2 = Color.FromName(colorName);
                // set the readable color name
                _swatch.Text = colorName;
            }

            // get the size of the color name string
            if (this.ElementTree != null)
            {
                SizeF size = SizeF.Empty;
                using (Graphics g = this.ElementTree.Control.CreateGraphics())
                {
                    size = g.MeasureString(_swatch.Text, this.Font);
                }
            
                // get the current column width
                int width = this.GridControl.Columns[this.ColumnIndex].Width;
                // set column width to use the column or measured 
                // color name string width, whichever is widest
                this.GridControl.Columns[this.ColumnIndex].Width = Math.Max(width, (int)size.Width);
            }
        }

        void _swatch_MouseDown(object sender, MouseEventArgs e)
        {
            // configure color dialog to only show web colors
            RadColorDialog colorDialog = new RadColorDialog();
            colorDialog.ColorDialogForm.ShowBasicColors = false;
            colorDialog.ColorDialogForm.ShowProfessionalColors = false;
            colorDialog.ColorDialogForm.ShowSystemColors = false;
            colorDialog.ColorDialogForm.ShowCustomColors = false;
            colorDialog.ColorDialogForm.ShowHEXColorValue = false;
            colorDialog.ColorDialogForm.AllowColorPickFromScreen = false;
            colorDialog.ColorDialogForm.AllowColorSaving = false;
            colorDialog.ColorDialogForm.WebTabHeading = "Swatch Colors";

            // set initial dialog color
            colorDialog.SelectedColor = _swatch.BackColor2;
            if (colorDialog.ShowDialog() == DialogResult.OK)
            {
                // set value to known color name of selected color
                this.Value = GetColorName(colorDialog.SelectedColor);
            }
        }

        public string GetColorName(Color color)
        {
            foreach (KnownColor knownColor in Enum.GetValues(typeof(KnownColor)))
            {
                // only look for web colors, not system colors
                if ((knownColor > KnownColor.Transparent) && (knownColor < KnownColor.ButtonFace))
                {
                    if (color.ToArgb() == Color.FromKnownColor(knownColor).ToArgb())
                    {
                        return knownColor.ToString();
                    }
                }
            }
            return KnownColor.Transparent.ToString();
        }

        public override bool IsCompatible(GridViewColumn data, object context)
        {
            return data is ColorSwatchColumn && context is GridDataRowElement;
        }

        protected override Type ThemeEffectiveType
        {
            get
            {
                return typeof(GridDataCellElement);
            }
        }
    }
}

File: telerik_ui_for_winforms_courseware_chm/Courseware/Grid/VB/RadGridView/CustomCells/ColorSwatchCellElement.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.Windows.Forms
Imports Telerik.WinControls
Imports Telerik.WinControls.UI


Public Class ColorSwatchCellElement
    Inherits GridDataCellElement
    Public Sub New(ByVal column As GridViewColumn, ByVal row As GridRowElement)
        MyBase.New(column, row)
    End Sub

    Private _swatch As LightVisualElement

    Protected Overloads Overrides Sub CreateChildElements()
        MyBase.CreateChildElements()

        _swatch = New LightVisualElement()
        _swatch.DrawFill = True
        _swatch.Font = New Font("Arial", 7.0F, FontStyle.Regular)
        _swatch.Shape = New RoundRectShape(5)
        _swatch.GradientStyle = GradientStyles.Linear
        _swatch.NumberOfColors = 2
        RemoveHandler _swatch.MouseDown, AddressOf _swatch_MouseDown
        AddHandler _swatch.MouseDown, AddressOf _swatch_MouseDown
        Me.Children.Add(_swatch)
    End Sub

    Protected Overrides Function ArrangeOverride(ByVal finalSize As SizeF) As SizeF
        MyBase.ArrangeOverride(finalSize)
        'size the element according to the cell size
        _swatch.Arrange(New RectangleF(3, 3, finalSize.Width - 6, finalSize.Height - 6))
        Return finalSize
    End Function

    Protected Overloads Overrides Sub SetContentCore(ByVal value As Object)
        ' blank out the cell's text, only show on swatch element
        Me.Text = ""
    End Sub

    Public Overloads Overrides Sub SetContent()
        MyBase.SetContent()

        ' set default properties
        _swatch.BackColor = Color.White
        _swatch.BackColor2 = Color.White
        _swatch.Text = String.Empty

        ' only process if there is a value and its not "Transparent"
        If (Not Me.Value.Equals(System.DBNull.Value)) AndAlso ((Not Me.Value.ToString().Equals("Transparent"))) Then
            Dim colorName As String = Me.Value.ToString()
            ' convert to a Color type and assign to background color
            _swatch.BackColor2 = Color.FromName(colorName)
            ' set the readable color name
            _swatch.Text = colorName
        End If

        If Me.ElementTree IsNot Nothing Then
            ' get the size of the color name string
            Dim size As SizeF = SizeF.Empty
            Using g As Graphics = Me.ElementTree.Control.CreateGraphics()
                size = g.MeasureString(_swatch.Text, Me.Font)
            End Using

            ' get the current column width
            Dim width As Integer = Me.GridControl.Columns(Me.ColumnIndex).Width
            ' set column width to use the column or measured 
            ' color name string width, whichever is widest
            Me.GridControl.Columns(Me.ColumnIndex).Width = Math.Max(width, CInt(Fix(size.Width)))
        End If
    End Sub

    Private Sub _swatch_MouseDown(ByVal sender As Object, ByVal e As MouseEventArgs)
        ' configure color dialog to only show web colors
        Dim colorDialog As RadColorDialog = New RadColorDialog()
        colorDialog.ColorDialogForm.ShowBasicColors = False
        colorDialog.ColorDialogForm.ShowProfessionalColors = False
        colorDialog.ColorDialogForm.ShowSystemColors = False
        colorDialog.ColorDialogForm.ShowCustomColors = False
        colorDialog.ColorDialogForm.ShowHEXColorValue = False
        colorDialog.ColorDialogForm.AllowColorPickFromScreen = False
        colorDialog.ColorDialogForm.AllowColorSaving = False
        colorDialog.ColorDialogForm.WebTabHeading = "Swatch Colors"

        ' set initial dialog color
        colorDialog.SelectedColor = _swatch.BackColor2
        If colorDialog.ShowDialog() = System.Windows.Forms.DialogResult.OK Then
            ' set value to known color name of selected color
            Me.Value = GetColorName(colorDialog.SelectedColor)
        End If
    End Sub

    Public Function GetColorName(ByVal color As Color) As String
        For Each knownColor As KnownColor In System.Enum.GetValues(GetType(KnownColor))
            ' only look for web colors, not system colors
            If (knownColor > knownColor.Transparent) AndAlso (knownColor < knownColor.ButtonFace) Then
                If color.ToArgb() = System.Drawing.Color.FromKnownColor(knownColor).ToArgb() Then
                    Return knownColor.ToString()
                End If
            End If
        Next knownColor
        Return KnownColor.Transparent.ToString()
    End Function

    Protected Overrides ReadOnly Property ThemeEffectiveType() As Type
        Get
            Return GetType(GridDataCellElement)
        End Get
    End Property

    Public Overrides Function IsCompatible(ByVal data As GridViewColumn, ByVal context As Object) As Boolean
        Return TypeOf data Is ColorSwatchColumn AndAlso TypeOf context Is GridDataRowElement
    End Function
End Class
